src/lib/datepicker/datepicker.spec.ts

```diff
@@ -1673,3 +1689,14 @@ class DatepickerWithEvents {
 class DatepickerOpeningOnFocus {
   @ViewChild(MatDatepicker) datepicker: MatDatepicker<Date>;
 }
+
+@Component({
+  template: `
+    <input [matDatepicker]="ch">
+    <mat-datepicker #ch></mat-datepicker>
+  `,
+})
+class DatepickerWithCustomHeader {
+  @ViewChild('ch') datepicker: MatDatepicker<Date>;
+  @ViewChild(MatDatepickerInput) datepickerInput: MatDatepickerInput<Date>;
+}
```

Comment:
@mmalerba 

I tried to add the custom header component as an input:

```typescript
@Component({
  template: `
    <input [matDatepicker]="ch">
    <mat-datepicker #ch [calendarHeaderComponent]="CustomHeaderForDatepicker"></mat-datepicker>
  `,
})
class DatepickerWithCustomHeader {
  @ViewChild('ch') datepicker: MatDatepicker<Date>;
  @ViewChild(MatDatepickerInput) datepickerInput: MatDatepickerInput<Date>;
}

@Component({
  template: `
  <div>Custom element</div>
  <mat-calendar-header></mat-calendar-header>
  `
})
export class CustomHeaderForDatepicker {
  constructor() {
  }
}
```

But I get:

> : Cannot determine the module for class CustomHeaderForDatepicker in /src/lib/datepicker/datepicker.spec.ts! Add CustomHeaderForDatepicker to the NgModule to fix it.

I remember that I had to add the custom header component to `entryComponents` for the demo page. How would that work for a test?

Reply:
Is this a possible solution https://stackoverflow.com/questions/41483841/providing-entrycomponents-for-a-testbed?

Reply:
You can modify the `createComponent` function to take an optional list of `entryComponents`. It already had optional lists for `imports` and `providers`

Reply:
`entryComponents` is not a member of the expected configuration object

```ts
static configureTestingModule(moduleDef: TestModuleMetadata): typeof TestBed;
```

```ts
export declare type TestModuleMetadata = {
    providers?: any[];
    declarations?: any[];
    imports?: any[];
    schemas?: Array<SchemaMetadata | any[]>;
    aotSummaries?: () => any[];
};
```

Is this the way to do it https://github.com/angular/angular/issues/10760#issuecomment-250522300 ?

Reply:
So the proposed solution would be 

```ts
// Creates a test component fixture.
  function createComponent(
    component: any,
    imports: any[] = [],
    providers: any[] = [],
    entryComponents: any[] = []):
    ComponentFixture<any> {

    TestBed.configureTestingModule({
      imports: [
        FormsModule,
        MatDatepickerModule,
        MatFormFieldModule,
        MatInputModule,
        NoopAnimationsModule,
        ReactiveFormsModule,
        ...imports
      ],
      providers,
      declarations: [component],
    });

    TestBed.overrideModule(BrowserDynamicTestingModule, {
      set: {
        entryComponents: [entryComponents]
      }
    }).compileComponents();

    return TestBed.createComponent(component);
  }

```

Reply:
I think maybe the fact that you're exporting the class is the problem. Did you try just 

```ts
class CustomHeaderForDatepicker { ... }
```

Reply:
@mmalerba 

done in https://github.com/angular/material2/pull/10856/commits/6aca9342888a4920fed71bb8f3d4e78b74584eb7

However, there are still some issues:

- tslint complains about `import {BrowserDynamicTestingModule} from '@angular/platform-browser-dynamic/testing';` -> missing in `rollup-gloabls.ts`
- the test is executed now but fails. It complains about a missing import. But if I add `CustomHeaderForDatepicker` as an import to `createComponent` it fails too.

I would be glad for any advice ;-)

Reply:
Ah, I forgot to add the custom header to `declarations` (https://stackoverflow.com/questions/44827999/component-is-not-part-of-any-ngmodule-or-the-module-has-not-been-imported-into-y?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa).

Done in https://github.com/angular/material2/pull/10856/commits/5ee44e78a0f30d9fc7a748d7ed912dad1bff758d Sorry!

lint problem should be fixed in https://github.com/angular/material2/pull/10856/commits/e189b239c46bff1b73901dadef5488d1e3d46779